src/Serilog/Settings/KeyValuePairs/KeyValuePairSettings.cs

```diff
@@ -135,12 +159,63 @@ select new
                 }
             }
         }
 
-        static void ApplyDirectives(List<IGrouping<string, ConfigurationMethodCall>> directives, IList<MethodInfo> configurationMethods, object loggerConfigMethod)
+        static IReadOnlyDictionary<string, LoggingLevelSwitch> ParseNamedLevelSwitchDeclarationDirectives(Dictionary<string, string> directives)
+        {
+            var matchLevelSwitchDeclarations = new Regex(LevelSwitchDeclarationDirectiveRegex);
+
+            var switchDeclarationDirectives = (from wt in directives
+                                               where matchLevelSwitchDeclarations.IsMatch(wt.Key)
+                                               let match = matchLevelSwitchDeclarations.Match(wt.Key)
+                                               select new
+                                               {
+                                                   SwitchName = match.Groups["switchName"].Value,
+                                                   InitialSwitchLevel = wt.Value
+                                               }).ToList();
+
+            var namedSwitches = new Dictionary<string, LoggingLevelSwitch>();
+            foreach (var switchDeclarationDirective in switchDeclarationDirectives)
+            {
+                LoggingLevelSwitch newSwitch;
+                if (string.IsNullOrEmpty(switchDeclarationDirective.InitialSwitchLevel))
+                {
+                    newSwitch = new LoggingLevelSwitch();
+                }
+                else
+                {
+                    var initialLevel = (LogEventLevel)SettingValueConversions.ConvertToType(switchDeclarationDirective.InitialSwitchLevel, typeof(LogEventLevel));
+                    newSwitch = new LoggingLevelSwitch(initialLevel);
+                }
+                namedSwitches[switchDeclarationDirective.SwitchName] = newSwitch;
+
+            }
+            return new ReadOnlyDictionary<string, LoggingLevelSwitch>(namedSwitches);
+        }
+
+        static LoggingLevelSwitch LookUpSwitchByNameOrThrow(string minimumLevelControlledByLevelSwitchName, IReadOnlyDictionary<string, LoggingLevelSwitch> declaredLevelSwitches)
+        {
+            if (!declaredLevelSwitches.ContainsKey(minimumLevelControlledByLevelSwitchName))
+            {
+                throw new InvalidOperationException($"No LoggingLevelSwitch has been declared with name \"{minimumLevelControlledByLevelSwitchName}\". You might be missing a key \"{LevelSwitchDirective}:$switchName\"");
+            }
+            var levelSwitch = declaredLevelSwitches[minimumLevelControlledByLevelSwitchName];
+            return levelSwitch;
+        }
+
+        static object ConvertOrLookupByName(string valueOrSwitchName, Type type, IReadOnlyDictionary<string, LoggingLevelSwitch> declaredSwitches)
+        {
+            if (type == typeof(LoggingLevelSwitch))
+            {
+                return LookUpSwitchByNameOrThrow(valueOrSwitchName, declaredSwitches);
```

Comment:
The naming scheme here isn't consistent - since `ConvertOrLookupByName()` could throw, I'd expect it to also carry the `OrThrow()` suffix if we're using that convention. Perhaps we could drop the `OrThrow()` from `LookUpSwitchByNameOrThrow()`, with the intention that we'd provide a `TryXxxx()` variant for the non-throwing version?